### PR TITLE
React 17 to 18 upgrade

### PR DESCRIPTION
**Pull request by dgboss**
- update React and related packages from v17 to v18
- remove code related to hot reload as it is supported natively in React 18
- fix unit tests broken by upgrade
- `React.StrictMode` doesn't play nice with keycloak-js with React 18 (you end up in an infinite authentication re-direct loop). Fix was to ensure the `AuthWrapper` component was not wrapped by `React.StrictMode`
- Fix duplicate of ASA map

# Test Links:
[Landing Page](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-3483.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: web/src/features/fba/components/map/FBAMap.tsx [hunk 1]
```diff
@@ -292,6 +292,9 @@ const FBAMap = (props: FBAMapProps) => {
     }
 
     setMap(mapObject)
+    return () => {
```
Comment (conbrad) [inline web/src/features/fba/components/map/FBAMap.tsx:295]: What problem does this solve?

File: web/src/features/moreCast2/components/GroupHeader.tsx [hunk 2]
```diff
@@ -64,7 +64,7 @@ const GroupHeader = ({ id, columns, weatherParam, handleShowHideChange }: GroupH
         <Stack sx={{ overflow: 'hidden' }}>
           <PopoverHeader>Choose Models to Display</PopoverHeader>
           <FormGroup>
-            {columns.map(column => {
+            {columns?.map(column => {
```
Comment (conbrad) [inline web/src/features/moreCast2/components/GroupHeader.tsx:67]: Looks like this isn't possibly undefined, do we need this change?

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx [hunk 3]
```diff
@@ -236,7 +236,7 @@ const TabbedDataGrid = ({ morecast2Rows, fromTo, setFromTo, fetchWeatherIndeterm
       handleClose: handleClose
     })
     // Provide default with all columns
-    const showHideColumnsUngroupedState = weatherModelColumns.map((column: GridColDef): ColumnVis => {
+    const showHideColumnsUngroupedState = weatherModelColumns?.map((column: GridColDef): ColumnVis => {
```
Comment (conbrad) [inline web/src/features/moreCast2/components/TabbedDataGrid.tsx:239]: Are these possibly undefined now? Maybe the react types changed?

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx [hunk 4]
```diff
@@ -284,7 +284,7 @@ const TabbedDataGrid = ({ morecast2Rows, fromTo, setFromTo, fetchWeatherIndeterm
 
   useEffect(() => {
     const newVisibleRows: MoreCast2Row[] = []
-    const stationCodes = selectedStations.map(station => station.station_code)
+    const stationCodes = selectedStations?.map(station => station.station_code)
```
Comment (conbrad) [inline web/src/features/moreCast2/components/TabbedDataGrid.tsx:287]: Are these possibly undefined now? Maybe the react types changed?

**Review by conbrad — APPROVED** [review]
Nice work!

File: web/src/features/fba/components/map/FBAMap.tsx (hunk 1 above)
Comment (dgboss) [reply web/src/features/fba/components/map/FBAMap.tsx:295]: For some reason with React 18, the 'double' rendering that happens when running a dev build led to two `FBAMaps` being rendered in ASA. This cleanup function effectively unmounts/removes the map from its parent div so we don't have two maps showing up.

**Review by brettedw — APPROVED** [review]
LGTM. The only confusing part was the part that you commented on with the double map rendering. Seems like a strange behaviour but I trust your fix 👌

File: web/src/features/moreCast2/components/GroupHeader.tsx (hunk 2 above)
Comment (dgboss) [reply web/src/features/moreCast2/components/GroupHeader.tsx:67]: The app works fine with the null check, but unit tests fail.

It looks like columns can be undefined when the app starts up. The culprit is:
`const [showHideColumnsModel, setShowHideColumnsModel] = useState<Record<string, ColumnVis[]>>({})`

We start with an empty object and down the chain we attempt to access the temp/rh/wind/precip properties of the empty object which don't yet exist which ends up with columns=undefined being passed in to the `GroupHeader` component. 

I might be able to fix it by initializing `showHideColumnsModel` in the `useState` hook.

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx (hunk 3 above)
Comment (dgboss) [reply web/src/features/moreCast2/components/TabbedDataGrid.tsx:239]: I was trying to debug unit tests which complained of .map being called in undefined and this got left behind as an artifact of my debugging. 

I'll revert the change. Thanks!

File: web/src/features/moreCast2/components/TabbedDataGrid.tsx (hunk 4 above)
Comment (dgboss) [reply web/src/features/moreCast2/components/TabbedDataGrid.tsx:287]: I was trying to debug unit tests which complained of .map being called in undefined and this got left behind as an artifact of my debugging.

I'll revert the change. Thanks!

File: web/src/features/moreCast2/components/GroupHeader.tsx (hunk 2 above)
Comment (dgboss) [reply web/src/features/moreCast2/components/GroupHeader.tsx:67]: Yep, looks like initialization in the useState hook fixes the problem. I've reverted the null check.

**Pull request merged**